cli/ssh_test.go

```diff
@@ -644,32 +646,36 @@ func TestSSH(t *testing.T) {
 			"8222:"+httpServer.Listener.Addr().String(),
 		)
 		clitest.SetupConfig(t, client, root)
-		pty := ptytest.New(t).Attach(inv)
-		inv.Stderr = pty.Output()
 
 		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
 		defer cancel()
 
 		cmdDone := tGo(t, func() {
 			err := inv.WithContext(ctx).Run()
-			assert.NoError(t, err, "ssh command failed")
+			// fails because we cancel context to close
+			assert.Error(t, err, "ssh command should fail")
 		})
 
-		// Agent is still starting
-		pty.ExpectMatch("Waiting")
-
-		_ = agenttest.New(t, client.URL, agentToken)
-		coderdtest.AwaitWorkspaceAgents(t, client, workspace.ID)
-
-		// Startup script has just finished
-		pty.ExpectMatch(startupScriptPattern)
-
-		// Download the test page
-		pty.WriteLine("curl localhost:8222")
-		pty.ExpectMatch("hello world")
+		require.Eventually(t, func() bool {
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, "http://localhost:8222/", nil)
+			if !assert.NoError(t, err) {
+				// true exits the loop.
+				return true
+			}
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				t.Logf("HTTP GET http://localhost:8222/ %s", err)
+				return false
+			}
+			defer resp.Body.Close()
+			body, err := io.ReadAll(resp.Body)
+			assert.NoError(t, err)
+			assert.EqualValues(t, "hello world", body)
+			return true
+		}, testutil.WaitLong, testutil.IntervalFast)
```

Comment:
The previous matcher waited for `i-am-ready` echoed by the startup script. It means that the agent didn't start yet, and `localhost:8222` is not available. I'm curious where is the trick in the new implementation. Is it a matter of `testutil.WaitLong`?

Reply:
Yeah, before we waited for `i-am-ready` and then connected.

Here, we just (re)try connecting in a loop until success or a timeout.

Reply:
Isn't it the same root cause? The agent is still not ready to connect. Are you just trying to reduce complexity and remove extra conditions?

Reply:
Before we used PTY/expect to wait until we thought the agent was connected and SSH remote forward was up, then we connected and verified the output.  The root cause of the flake is that we didn't always get the startup script log because of the way the `ssh` command works.  It detects that a startup script is running, then asks to follow logs, but the log may have already been dispatched and we miss it, causing the expect to hang.

In the new implementation we do away with PTY/expect entirely, and thus dodge the root cause from before.  To deal with the fact that we no longer get an explicit signal that the remote forward is up, we just retry it multiple times until it comes up.

Reply:
> The root cause of the flake is that we didn't always get the startup script log because of the way the ssh command works.

Ok, that's the part I wasn't aware of. Thanks!